### PR TITLE
Adding support for explain in rrf

### PR DESCRIPTION
**Pull request by pmpailis**
In this PR we add support for explaining results and scores in RRF queries. To do this, this PR makes use of the `ExplainPhase` fetch sub-phase that is executed on the shards when `explain=true`. To achieve this a new `explainHit` method is specified for each ranker, so that they take a `RankDoc` and provide explanations on how the final score and ranks have been computed.

Since this takes place on the shards, we also have to serialize and pass the `RankDoc` objects that are associated with each shard during the `FetchSearchPhase` execution. This is enabled **iff** both `explain` and a `RankBuilder` are defined in the request, otherwise there is no impact at all (i.e. a null object is to be serialized and sent).

We also aim to match the queries executed through the `RankBuilder` with any `_name` associated with them, so that the output can be easier to interpret (if no `_name` is provided for a query, then we just make use of the query's index when executed). 

Example RRF explain output:
```
{
    "hits":
    [
        {
            "_index": "example-index",
            "_id": "3",
            "_score": 0.8333334,
            "_rank": 1,
            "_explanation":
            {
                "value": 0.8333334,                                                                                                                                               
                "description": "rrf score: [0.8333334] computed for initial ranks [2, 1] with rankConstant: [1] as sum of [1 / (rank + rankConstant)] for each query",            
                "details":                                                                                                                                                        
                [
                    {
                        "value": 2,                                                                                                                                               
                        "description": "rrf score: [0.33333334], for rank [2] in query at index [0] computed as [1 / (2 + 1]), for matching query with score: ",
                        "details":                                                                                                                                                
                        [
                            {
                                "value": 0.15876243,
                                "description": "weight(text:rrf in 0) [PerFieldSimilarity], result of:",
                                "details":
                                [
                                    ...
                                ]
                            }
                        ]
                    },
                    {
                        "value": 1,                                                                                                                                              
                        "description": "rrf score: [0.5], for rank [1] in query ["my_knn_query"] computed as [1 / (1 + 1]), for matching query with score: ",
                        "details":
                        [
                            {
                                "value": 1,
                                "description": "within top k documents",
                                "details":
                                []
                            }
                        ]
                    }
                ]
            }
        }
        ...
    ]
}
```

### CODE REVIEW

File: docs/reference/search/rrf.asciidoc [hunk 1]
```diff
@@ -516,6 +515,157 @@ We end with `_id: 3` as `_rank: 1`, `_id: 2` as `_rank: 2`, and
...
+// NOTCONSOLE
+
+<1> the final RRF score for document with `_id=3`
+<2> a description on how this score was computed based on the ranks of this document in each individual query
+<3> details on how the RRF score was computed for each of the queries
+<4> the `value` heres specifies the `rank` of this document in the specific query
+<5> standard `explain` output of the underlying query, describing matching terms and weights
+<6> the `value` heres specifies the `rank` of this document for the second (`knn`) query
+
+In addition to the above, explain in RRF also supports *named queries* using the `_name` parameter.
```
Comment (benwtrent) [inline docs/reference/search/rrf.asciidoc:584]: link the named query docs. `<<named-queries,named queries>>`

File: docs/reference/search/rrf.asciidoc [hunk 2]
```diff
@@ -516,6 +515,157 @@ We end with `_id: 3` as `_rank: 1`, `_id: 2` as `_rank: 2`, and
...
+This however is not available for retrievers yet, so it is only supported for RRF through the use of  `sub_searches`,
```
Comment (benwtrent) [inline docs/reference/search/rrf.asciidoc:585]: Do you mean that `retrievers` don't support `_name` yet? Or that they don't support `explain`?

File: docs/reference/search/rrf.asciidoc [hunk 3]
```diff
@@ -516,6 +515,157 @@ We end with `_id: 3` as `_rank: 1`, `_id: 2` as `_rank: 2`, and
...
+and for top-level `query` and `knn` elements for now. Using named queries allows for easier and more intuitive understanding of the RRF score computation,
+especially when dealing with multiple queries. So, for the equivalent of the above example, we would now have:
+
+[source,js]
+----
+GET example-index/_search
+{
+    "sub_searches": [
```
Comment (benwtrent) [inline docs/reference/search/rrf.asciidoc:593]: sub_searches is going away right with retrievers? We should not add any new documentation or features using it. We need to focus on retrievers.

File: server/src/main/java/org/elasticsearch/search/fetch/subphase/ExplainPhase.java [hunk 4]
```diff
@@ -40,6 +40,9 @@ public void process(HitContext hitContext) throws IOException {
                 for (RescoreContext rescore : context.rescore()) {
                     explanation = rescore.rescorer().explain(topLevelDocId, context.searcher(), rescore, explanation);
                 }
+                if (context.rankBuilder() != null) {
+                    explanation = context.rankBuilder().explainHit(explanation, hitContext.rankDoc(), context.queryNames());
```
Comment (benwtrent) [inline server/src/main/java/org/elasticsearch/search/fetch/subphase/ExplainPhase.java:44]: For every hit, `context.queryNames()` makes this call
```
searchContext.request().source().subSearches().stream().map(x -> x.getQueryBuilder().queryName()).toList();
```

This is unnecessarily expensive. Think you could statically set it in the processor and calculate it once?

File: test/framework/src/main/java/org/elasticsearch/search/rank/TestRankDoc.java [hunk 5]
```diff
@@ -37,4 +37,9 @@ public boolean doEquals(RankDoc rd) {
     public int doHashCode() {
         return 0;
     }
+
+    @Override
+    public String getWriteableName() {
+        return "test-rank-doc";
```
Comment (benwtrent) [inline test/framework/src/main/java/org/elasticsearch/search/rank/TestRankDoc.java:43]: prefer `_` everywhere for names like this.

File: server/src/main/java/org/elasticsearch/search/rank/RankDoc.java [hunk 6]
```diff
@@ -21,7 +21,7 @@
  * Subclasses should extend this with additional information
  * required for their global ranking method.
  */
-public abstract class RankDoc extends ScoreDoc implements Writeable {
+public abstract class RankDoc extends ScoreDoc implements NamedWriteable {
```
Comment (benwtrent) [inline server/src/main/java/org/elasticsearch/search/rank/RankDoc.java:24]: How do you get away with this without adjusting all current serialization for `RankDoc`? 

Switching from writeable to namewriteable might break some things? Maybe I missed it.

File: x-pack/plugin/rank-rrf/src/yamlRestTest/resources/rest-api-spec/test/rrf/500_rrf_explain.yml [hunk 7]
```diff
@@ -0,0 +1,386 @@
+setup:
```
Comment (benwtrent) [inline x-pack/plugin/rank-rrf/src/yamlRestTest/resources/rest-api-spec/test/rrf/500_rrf_explain.yml:1]: No tests with retrievers?

File: server/src/main/java/org/elasticsearch/search/fetch/subphase/ExplainPhase.java (hunk 4 above)
Comment (pmpailis) [reply server/src/main/java/org/elasticsearch/search/fetch/subphase/ExplainPhase.java:44]: Updated!

File: server/src/main/java/org/elasticsearch/search/rank/RankDoc.java (hunk 6 above)
Comment (pmpailis) [reply server/src/main/java/org/elasticsearch/search/rank/RankDoc.java:24]: The only concrete implementations up until now where the `RRFRankDoc` and `TestRankDoc` (used in tests). In both cases I added the impl for the `getWriteableName` method and added this to `RRFRankPlugin`'s exported `getNamedWriteables` . 

The issue was that up until now we explicitly mapped to a `RRFRankDoc` in `RRFRankShardResult`, however now the `fetch` phase could in theory accept any arbitrary concrete implementation of the `RankDoc` class.  So this is why I had to update to a `NamedWriteable`, to make use of (in `RankDocShardInfo`) 
```
        rankDocs = in.readMap(StreamInput::readVInt, v -> v.readNamedWriteable(RankDoc.class));

```
Is there any additional serialization needed that I missed? :/

File: docs/reference/search/rrf.asciidoc (hunk 2 above)
Comment (pmpailis) [reply docs/reference/search/rrf.asciidoc:585]: IIUC `standard` does support `_name`, whereas `knn` retriever doesn't. For example the following: 
```

GET my_test_index/_search
{
    "retriever": {
        "knn": {
            "field": "my_vector_byte",
            "query_vector": [
                1,
                -2,
                0
            ],
            "num_candidates": 10,
            "k": 10,
            "_name": "my_knn_query"
        }
    }
}
```
returns -> 
```
{
    "error": {
        "root_cause": [
            {
                "type": "x_content_parse_exception",
                "reason": "[12:13] [knn] unknown field [_name]"
            }
        ],
        "type": "x_content_parse_exception",
        "reason": "[12:13] [knn] unknown field [_name]"
    },
    "status": 400
}
```
while for the `standard` retriever, it is parsed w/o any issues: 
```
GET my_test_index/_search
{
    "retriever": {
        "standard": {
            "query": {
                "term": {
                    "body": {
                        "value": "some_value",
                        "_name": "my_term_query"
                    }
                }
            }
        }
    }
}
```
I think that this is because for the `query` value of the `Standard` retriever, we make use of the `AbstractQueryBuilder` which supports `_name`, whereas for `knn` we parse directly all the top level params. 
I was planning on addressing this, but on a separate PR to avoid overcomplicating this one.

File: x-pack/plugin/rank-rrf/src/yamlRestTest/resources/rest-api-spec/test/rrf/500_rrf_explain.yml (hunk 7 above)
Comment (pmpailis) [reply x-pack/plugin/rank-rrf/src/yamlRestTest/resources/rest-api-spec/test/rrf/500_rrf_explain.yml:1]: Yeap, for some reasons I did not include any :/ Will add some similar tests but written using retrievers :)

File: docs/reference/search/rrf.asciidoc (hunk 2 above)
Comment (pmpailis) [reply docs/reference/search/rrf.asciidoc]: Added a first take in dada6f6d22a3ccd40bf47d02ca473e0fbcd0016a by adding support for parsing `_name` in the base `RetrieverBuilder` and exposing it for `knn` (which is a top-level query element).

File: x-pack/plugin/rank-rrf/src/main/java/org/elasticsearch/xpack/rank/rrf/RRFRankBuilder.java [hunk 8]
```diff
@@ -103,6 +106,62 @@ public QueryPhaseRankCoordinatorContext buildQueryPhaseCoordinatorContext(int si
         return new RRFQueryPhaseRankCoordinatorContext(size, from, rankWindowSize(), rankConstant);
     }
 
+    @Override
+    public Explanation explainHit(Explanation baseExplanation, RankDoc rankDoc, List<String> queryNames) {
```
Comment (benwtrent) [inline x-pack/plugin/rank-rrf/src/main/java/org/elasticsearch/xpack/rank/rrf/RRFRankBuilder.java:110]: This is so frustrating. The parser only support for retrievers is really hampering us from building this further into anything logical.

Really, all the explanation stuff SHOULD live in retrievers. Continuing to build on this bad interface is frustrating. There is no knowledge of relationship, everything is flattened out.

File: server/src/main/java/org/elasticsearch/search/rank/RankDoc.java (hunk 6 above)
Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/search/rank/RankDoc.java:24]: > Is there any additional serialization needed that I missed? :/

I don't know. The retrievers and rank code has evolved outside of my purview for a long time.

File: x-pack/plugin/rank-rrf/src/main/java/org/elasticsearch/xpack/rank/rrf/RRFRankBuilder.java (hunk 8 above)
Comment (pmpailis) [reply x-pack/plugin/rank-rrf/src/main/java/org/elasticsearch/xpack/rank/rrf/RRFRankBuilder.java:110]: +1 on this. I would too opt to have this on the `retriever` level, but unfortunately that's not an option.. And this gets even more frustrating when you want to build nested things, as the concept of "retriever" disappears right after parsing.. So any nested sub-retrievers, working on different outputs etc, is a bit of a pain to work with :(

File: x-pack/plugin/rank-rrf/src/yamlRestTest/resources/rest-api-spec/test/rrf/500_rrf_explain.yml (hunk 7 above)
Comment (pmpailis) [reply x-pack/plugin/rank-rrf/src/yamlRestTest/resources/rest-api-spec/test/rrf/500_rrf_explain.yml]: Added :)

File: x-pack/plugin/rank-rrf/src/main/java/org/elasticsearch/xpack/rank/rrf/RRFRankBuilder.java (hunk 8 above)
Comment (benwtrent) [reply x-pack/plugin/rank-rrf/src/main/java/org/elasticsearch/xpack/rank/rrf/RRFRankBuilder.java:110]: @pmpailis trying to figure out how to do explain for any nested thing will be rough. The only way to do this is if when we flatten it out, the original order is respected, then it turns into a iteration problem, not a recursive one.

File: server/src/main/java/org/elasticsearch/search/SearchModule.java [hunk 9]
```diff
@@ -1064,8 +1064,8 @@ private void registerFetchSubPhase(FetchSubPhase subPhase) {
     }
 
     private void registerRetrieverParsers(List<SearchPlugin> plugins) {
-        registerRetriever(new RetrieverSpec<>(StandardRetrieverBuilder.NAME, StandardRetrieverBuilder::fromXContent));
-        registerRetriever(new RetrieverSpec<>(KnnRetrieverBuilder.NAME, KnnRetrieverBuilder::fromXContent));
+        registerRetriever(new RetrieverSpec<>(StandardRetrieverBuilder.TYPE, StandardRetrieverBuilder::fromXContent));
+        registerRetriever(new RetrieverSpec<>(KnnRetrieverBuilder.TYPE, KnnRetrieverBuilder::fromXContent));
```
Comment (benwtrent) [inline server/src/main/java/org/elasticsearch/search/SearchModule.java:1068]: Honestly, `NAME` is preferred. Why did you add all this churn to switch to `TYPE`? These are `namedWriteables` and their serialization is nominally based.

Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/search/SearchModule.java:1068]: I see that these aren't writeables, but still, parsing via xcontent is called ` p.namedObject`. I don't see a reason for this switch.

Comment (pmpailis) [reply server/src/main/java/org/elasticsearch/search/SearchModule.java:1068]: This was changed in https://github.com/elastic/elasticsearch/commit/dada6f6d22a3ccd40bf47d02ca473e0fbcd0016a to avoid confusion, as I introduced a new `name` field for supporting `_name` param natively in retrievers (as a first attempt).

Comment (benwtrent) [reply server/src/main/java/org/elasticsearch/search/SearchModule.java:1068]: @pmpailis retrieverName is fine + Javadocs is fine. I don't think we should add extra churn switching from `NAME` to `TYPE` to indicate the named serializable information

**Review by benwtrent — APPROVED** [review]

**Comment by pmpailis** [comment]
@elasticmachine update branch

**Comment by pmpailis** [comment]
@elasticmachine update branch

**Comment by pmpailis** [comment]
@elasticmachine update branch

**Comment by pmpailis** [comment]
@elasticmachine update branch
